src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyStatus.cs

```diff
@@ -0,0 +1,299 @@
+                return false;
+            }
+
+            if (MaxConcurrentExecutionsSinceLastAdjustment < CurrentConcurrency)
+            {
+                // We only want to increase if we're fully utilizing our current concurrency level.
+                // E.g. if we increased to a high concurrency level, then events slowed to a trickle,
+                // we wouldn't want to keep increasing.
+                return false;
+            }
+
+            // a max parallelism of -1 indicates unbounded
+            return maxDegreeOfParallelism == -1 || CurrentConcurrency < maxDegreeOfParallelism;
+        }
+
+        internal int GetLatencyAdjustedInterval(int minIntervalMS, int defaultIntervalMS, int latencyMultiplier)
+        {
+            int resultInterval = defaultIntervalMS;
+
+            if (InvocationsSinceLastAdjustment > 0)
+            {
+                // Compute based on function latency, so faster functions can adjust more often, allowing their concurrency
+                // to "break away" from longer running, heavier functions.
+                // While a longer running function might not actually be a problem (e.g. might be I/O bound),
+                // this latency based prioritization is still beneficial. It allows fast functions to give
+                // us fast feedback. Worst case, for long running functions we just revert back to the default
+                // interval. So this is an optimization applied when possible.
+                int avgInvocationLatencyMs = (int)TotalInvocationTimeSinceLastAdjustmentMs / InvocationsSinceLastAdjustment;
+                int computedInterval = minIntervalMS + latencyMultiplier * avgInvocationLatencyMs;
+                resultInterval = Math.Min(computedInterval, defaultIntervalMS);
+            }
+
+            return resultInterval;
+        }
+
+        internal void IncreaseConcurrency()
+        {
+            int delta = GetNextAdjustment(1);
+            AdjustConcurrency(delta);
+        }
+
+        internal void DecreaseConcurrency()
+        {
+            int delta = GetNextAdjustment(-1);
+            AdjustConcurrency(-1 * delta);
+        }
+
+        internal void FunctionStarted()
+        {
+            lock (_syncLock)
+            {
+                OutstandingInvocations++;
+
+                if (OutstandingInvocations > MaxConcurrentExecutionsSinceLastAdjustment)
+                {
+                    // record the high water mark for utilized concurrency this interval
+                    MaxConcurrentExecutionsSinceLastAdjustment = OutstandingInvocations;
+                }
+            }
+        }
+
+        internal void FunctionCompleted(TimeSpan latency)
+        {
+            lock (_syncLock)
+            {
+                TotalInvocationTimeSinceLastAdjustmentMs += latency.TotalMilliseconds;
+                InvocationsSinceLastAdjustment++;
+                OutstandingInvocations--;
+            }
+        }
```

Comment:
The problem is that values like TotalInvocationTimeSinceLastAdjustmentMs and InvocationsSinceLastAdjustment need to be adjusted together atomically, since they're used elsewhere to make calculations (e.g. invocation latency), so the values need to be consistent.

Because the operations under the lock here are very fast, lock contention should be very low.